### ISSUE TITLE
Non differentiable points for freeform

### ISSUE THREAD

**Issue by brandonLi8** [phetsims/calculus-grapher#28]
From the design doc:

>Changes to the model: Graphs will no longer have extreme “jumps” at nondifferentiable points in f(x), but will have open circles instead. This will affect three different functions:  (hill, line, parabola).

However, playing around in the flash sim and in master, I was able to produce cusps (nondifferentiable) points with the freeform mode:

![image](https://user-images.githubusercontent.com/42391580/87201017-ffd3d600-c2ba-11ea-9c30-5190c47f615d.png)

@amanda-phet should there also be open circles (for the derivative) for the "sharp" points when manipulating the curve in "freeform" mode?

**Comment by amanda-phet** [comment]
Hm... I am OK with these sharp points if you feel like the model can handle putting open circles there in the derivative. If it's too complicated, we could force the curve to be smooth. 

You can't make a vertical line with the freeform tool, right? I want that to be avoided at all costs. It seems like there is a slight slope in your screenshot.

**Comment by brandonLi8** [comment]
Thanks @amanda-phet.

> You can't make a vertical line with the freeform tool, right? I want that to be avoided at all costs. It seems like there is a slight slope in your screenshot.

That's correct. Vertical lines are impossible for both flash and html5 (since there is only one point per x value of the curve partition).


> Hm... I am OK with these sharp points if you feel like the model can handle putting open circles there in the derivative. If it's too complicated, we could force the curve to be smooth.

OK. I'll start with trying to compute cusps in the derivative (by considering the slope on the left AND the right). If it gets too complicated, I'll report back.

**Comment by brandonLi8** [comment]
I went ahead and gave this a shot.

I attempted this by considering the slope of the secant lines from the left and the right. If they were approximately equal (differs by less than the `?derivativeThreshold` query parameter, which defaults to 0.65), the derivative exists and is the average of the slopes. Otherwise, the derivative doesn't exist.

Unfortunately, from my own testing, this solution doesn't work for all scenarios.

For triangle, it works well (I haven't added the circles yet):
![image](https://user-images.githubusercontent.com/42391580/87479093-18f4c380-c5e8-11ea-8af8-a7ba80b252c0.png)

However, for things like pedestal, where the derivative changes by a considerable amount but still is differentiable, the cusp detection is off:
![image](https://user-images.githubusercontent.com/42391580/87479069-0c706b00-c5e8-11ea-8f1c-727f527e3c80.png)

When I increase the `?derivativeThreshold`, it sometimes doesn't detect cusps for triangle.

**Comment by brandonLi8** [comment]
I am now convinced that this solution will not work.

I did some initial searching for solutions and found https://www.mathworks.com/matlabcentral/answers/57194-how-to-find-the-sharp-turn-in-a-2d-line-curve#answer_69185, which suggests to use curvature to detect cusps.

Trying this now.

**Comment by brandonLi8** [comment]
In the commit above, I tried detecting the curvature of 3 adjacent points. If the curvature was more than `?derivativeThreshold` (default now 12) then the point was considered a cusp. However, this yielded similar results to the screenshots above. I also added red dots on the original curve where the cusps were detected. No matter how I adjusted the query parameters there were some scenarios where cups where detected where the shouldn't have been and not detected when they should have been.
